Python: Serialization/Deserialization for types and partition values (port of conversions.py from python_legacy)

samredai opened the pull request:
This adds functionality for converting a value to bytes and vice-versa augmented by the `PrimitiveType` subclass provided.

Functionally this is a direct port of [conversion.py](https://github.com/apache/iceberg/blob/master/python_legacy/iceberg/api/types/conversions.py) in the python legacy library. The main changes are:
1. I replaced the large dictionaries containing `<type>: <lambda function>` pairs with [@singledispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch)
2. I removed the [TypeID](https://github.com/apache/iceberg/blob/master/python_legacy/iceberg/api/types/type.py#L24) enum and instead just used the types directly.

Overall this is more lines of source code but I'll summarize my reasonings for the above two changes:

1. I believe the intention behind the lambda maps was to allow for generic `from_...` and `to_...` functions that contain dynamic logic based on the argument type, similar to constructor overloading in Java. I feel that using `@singledispatch` achieves that more directly by creating a generic function and then registering functions based on the first argument's type. This also allows stacking the register decorators for types that share the same logic (such as `IntegerType` and `DateType`) instead of defining that logic multiple times. Also it's a stable feature added as early as python 3.4 so I believe it's future-proof.
2. The `TypeID` enum in the legacy python client carries some information in the enum value, for example `BOOLEAN = {"java_class": "Boolean.class", "python_class": bool, "id": 1}`. These dictionary values weren't used in `conversions.py` but it looks like they're used in other places such as the partition spec. I removed it because this also felt like it may be adding some unnecessary indirection. If anyone feels this should be put back I'm ok with that but I would suggest that we see where the relationships in these dictionaries are required and localize these mappings to that area of the source code (maybe with a simple `PY_TO_JAVA_CLASS = {...}` object and a `get_class_id(...)` function).

samredai commented:
cc: @cccs-eric @dramaticlly @CircArgs @rdblue @jun-he @cabhishek @TGooch44

dramaticlly commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(type_, partition_value_as_str: str):
```
hey @samredai , I appreciate your introduction of this dynamic overloading. I tried to adopt it in #4262 and looks like mypy is not happy about it as the method `_` now is being redefined multiple times with error code `no-redef`, do you know if we need to configure anything in mypy.ini or tox.ini to work around this issue?

dramaticlly replied:
Basically this is what I saw

```
(venv) ➜  python git:(master) ✗ mypy --no-implicit-optional --show-error-codes  --config tox.ini src
tox.ini: No [mypy] section in config file
src/iceberg/expression/literals.py:27: error: Library stubs not installed for "pytz" (or incompatible with Python 3.8)  [import]
src/iceberg/expression/literals.py:27: note: Hint: "python3 -m pip install types-pytz"
src/iceberg/expression/literals.py:27: note: (or run "mypy --install-types" to install all missing stub packages)
src/iceberg/expression/literals.py:27: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
src/iceberg/expression/literals.py:109: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:119: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:129: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:134: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:139: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:144: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:149: error: Name "_of" already defined on line 104  [no-redef]
```

samredai replied:
It looks like you're using your locally installed version of mypy which I'm guessing is earlier than version `0.920` which is when singledispatch was added to the whitelist. The `tox` suite shouldn't run into this since it uses the latest version of mypy (currently `0.931`).

Also although `@singledispatch` was added in python 3.4, `@singledispatchmethod` wasn't added until python 3.8. Fortunately it was a very small addition and it was back-ported to python 3.7, so you'll just have to add `singledispatch` to the `deps` in the tox.ini file and the pyproject.toml file.

kbendick commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+@from_bytes.register(DecimalType)
+def _(type_, b: bytes):
+    return Decimal(int.from_bytes(b, "big", signed=True) * 10**-type_.scale).quantize(
+        Decimal("." + "".join(["0" for i in range(1, type_.scale)]) + "1"), rounding=ROUND_HALF_UP
```
Question: I saw in another PR some assertions on the python type major version (checking for 3).

Are we supporting Python 2.7? If so, do we need to be concerned about differences between `range` semantics between the two?

samredai replied:
No we're definitely not supporting Python 2 since it's been officially deprecated for quite some time. Currently we're aiming to support 3.7+.

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+
+@from_partition_value_to_py.register(StringType)
+def _(type_, partition_value_as_str: str):
+    return partition_value_as_str
```
does this need to check for `partition_value_as_str == HIVE_NULL`?

samredai replied:
Yes I think you're right. I believe when a hive catalog is used, `"__HIVE_DEFAULT_PARTITION__"` can be the value for records with nulls in the partition column, regardless of the partition column type. I'm going to move this check out to it's own function and perform this check regardless of the partition type.

emkornfield replied:
I think this function might be the only other place that can actually be dispatched based on type.

emkornfield replied:
Also, to be clear, I'm not super familiar with Hive partitioning, so this really was a question which I don't know the answer to.

samredai replied:
It looks like in the java client this is really just for tests. I'm going to exclude this here but if it ends up being required it makes more sense to add it as part of the hive catalog implementation.

samredai replied:
Removed this.

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+
+@to_bytes.register(LongType)
+@to_bytes.register(TimeType)
+@to_bytes.register(TimestampType)
```
Is accepting values as `datetime.time` or `datetime.datetime` something that is handled at a higher level?  Or considered out of scope for the library, or something that wouldn't apply here?

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+@to_bytes.register(TimeType)
+@to_bytes.register(TimestampType)
+@to_bytes.register(TimestamptzType)
+def _(type_, value: Any) -> bytes:
```
I'm not sure if it works with `@singledispatch` but does it make sense to narrow the type of assumed values?

samredai replied:
Yes that should work, `@singledispatch` is only concerned with the first args type. I'll narrow the type hints! 👍

samredai replied:
Updated the type hints!

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+@to_bytes.register(BinaryType)
+@to_bytes.register(FixedType)
+def _(type_, value: Any) -> bytes:
+    return value
```
my understanding is type checking is only compile time checks.  It might make sense to validate the type of value here is actually bytes to provide a clearer error message to users.

samredai replied:
If we narrow the `value` argument's type hint and include type hints for the higher level objects, that should implicitly provide robust mypy validation that `value` is of type bytes if `type_` is `BinaryType` or `FixedType`. If we do need to validate the value types I believe that would be better in the higher level objects (since `foo.type=BinaryType` and `foo.value=<not bytes>` would indicate incorrect logic somewhere).

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+
+@to_bytes.register(DecimalType)
+def _(type_, value: Any) -> bytes:
+    scale = abs(value.as_tuple().exponent)
```
does there need to be conversion from the implied scale of value and the scale specified in type?

samredai replied:
I'm not 100% sure on the requirement here. If a value is provided with a higher scale than the defined type, it feels like just converting it might have implications when lower/upper bound values are deserialized later. I'll need to dig in a bit more here.

emkornfield replied:
I think given that fact that we expect this to be used from literals, it seems like the normalize should probably happen at that level, and here there should be an assert that the scales are equal.

samredai replied:
Added a check and raise for mismatched precision and/or scale.

emkornfield commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,181 @@
+)
+def test_from_partition_value_to_py_approximated(type_, partition_value_as_str, expected_result):
+    """Test approximate equality when converting a partition value to a python built-in"""
+    assert conversions.from_partition_value_to_py(type_, partition_value_as_str) == pytest.approx(expected_result)
```
could this be made non-approx if Decimal("123.45") was used?

samredai replied:
You're absolutely right, I'll change that and remove this approximation test.

samredai replied:
Updated!

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+
+@from_bytes.register(DecimalType)
+def _(type_, b: bytes):
+    return Decimal(int.from_bytes(b, "big", signed=True) * 10**-type_.scale).quantize(
```
Does this work for something like Decimal(38, 37)?  It seems that using the [tuple constructor](https://docs.python.org/3/library/decimal.html#decimal-objects) would be more robust?  Or `Decimal(int.from_bytes(...)) * Decimal(10 ** -type.scale)` (a context might need to be passed in for this option)?

samredai replied:
Ah right, this would run into an issue there. I like the idea to use the tuple constructor and provide a context. Something like:
```py
@from_bytes.register(DecimalType)
def _(type_, b: bytes):
    integer_representation = int.from_bytes(b, "big", signed=True)
    sign = 1 if integer_representation < 0 else 0
    integer_tuple = tuple(int(i) for i in str(abs(integer_representation)))
    return Decimal(value=(sign, integer_tuple, -type_.scale), context=Context(prec=type_.precision, rounding=ROUND_HALF_UP))
```

emkornfield replied:
I think the separate Decimal option might be more slightly more efficient because it doesn't require the string coversion, if it works.

samredai replied:
I see, ok let me try that!

samredai replied:
Updated the `to_bytes(...)` and `from_bytes(...)` methods dispatched for `DecimalType`. I use the decimal option for from_bytes and I added the scale and precision checks in to_bytes.

jun-he commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,251 @@
+
+
+@singledispatch
+def from_partition_value_to_py(type_, partition_value_as_str: str):
```
BTW, also is it feasible to put type annotation for `type_`?

samredai replied:
Agreed, I'll rename `type_` to `primitive_type`, that's much clearer. For adding the type annotation, this should be feasible in that I don't think it will break the single dispatch decorator. Are you thinking we hint using the base primitive type class like `def from_partition_value_to_py(type_: PrimitiveType, partition_value_as_str: str):`?

samredai replied:
Updated! Thanks @jun-he, this looks much cleaner.

emkornfield commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,181 @@
+        (FixedType(3), b"foo", b"foo"),
+        (BinaryType(), b"foo", b"foo"),
+        (DecimalType(5, 2), b"\x30\x39", Decimal(123.45).quantize(Decimal(".01"))),
+        (DecimalType(5, 4), b"\x00\x12\xd6\x87", Decimal(123.4567).quantize(Decimal(".0001"))),
```
similar comment here, I think it might be clearer to use string constructor on Decimal.

samredai replied:
Converted this everywhere in the tests so quantize isn't used anymore.

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,261 @@
+
+    if value_scale != primitive_type.scale:
+        raise ValueError(f"Cannot serialize value, scale of value does not match type {primitive_type}: {value_scale}")
+    elif value_precision != primitive_type.precision:
```
Isn't it valid for value to have less digits then precision.

e.g. how is something like "1234.0" handled for Decimal(6, 1)?

samredai replied:
Yes that's totally valid and the spec explicitly says that the precision for a decimal column can be widened. I'll change this to just a check for precision greater than specified by the type. Thanks for catching that!

samredai replied:
Fixed!

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,261 @@
+    integer_representation = int.from_bytes(b, "big", signed=True)
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        return Decimal(integer_representation) * Decimal(10) ** Decimal(-primitive_type.scale)
```
```suggestion
        return Decimal(integer_representation) * (Decimal(10) ** Decimal(-primitive_type.scale))
```
Lets try to avoid relying on order of operations by python.

emkornfield commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,288 @@
+        (
+            DecimalType(7, 2),
+            Decimal("1234.56"),
+            "Cannot serialize value, precision of value does not match type decimal(7, 2): 6",
```
To follow-up on my question on the implementation. How can a user specifying this value?  It seems like it should only be an error if precision is exceeded (as you test for below).

samredai replied:
Replied [here](https://github.com/apache/iceberg/pull/4263#discussion_r822026954) and fixed!

emkornfield commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,288 @@
+@pytest.mark.parametrize(
+    "primitive_type, b, result",
+    [
+        (FloatType(), b"\x19\x04\x9e?", 1.2345),
```
do these fail if they aren't approximated.  I would think that serializion to and from bytes should be deterministic and not subject to numeric inaccuracies.

samredai replied:
I kept this consistent with the python legacy client which uses the built-in float and double types and just [tests](https://github.com/apache/iceberg/blob/master/python_legacy/tests/api/test_conversions.py#L48#L51) for approximation. According to the Iceberg spec though the float and double primitive types are supposed to be 32-bit and 64-bit IEEE floating points, so the easiest way I think is to use numpy here which was proposed by @CircArgs at some point.

@dramaticlly this would also require that numpy is used in PR #4262 for `FloatLiteral` and `DoubleLiteral`.

emkornfield replied:
Could you elaborate on why is numpy needed here?  

I expect python equality to work, since this is a deterministic round-trip of the data (is it something in the testing framework):
```
>>> 1.2345 == 1.2345
True
```

samredai replied:
Yes equality works but the imprecision surfaces when converting the float to a c struct and back:
```py
import struct

b = struct.pack("<f", 1.2345)
value = struct.unpack("<f", b)  # 1.2345000505447388
```
and `1.2345 == 1.2345000505447388` returns `False`

emkornfield replied:
Interesting, I dug bit further and turned up [this old bug report](https://bugs.python.org/issue4114) which has a couple of solutions to round-trip deterministically (but involves truncation.  I think this is only an issue for Floats and Doubles should round-trip deterministically. I assume the byte values are taken from Java?

samredai replied:
There will eventually be a write path from python so the byte values stored in the metadata could potentially be from python or Java. But I'd expect today it's almost entirely byte values generated in Java. As the python library matures we'll eventually have some cross-language tests to make sure the behaviors are synced (to_bytes from Java -> from_bytes from python, for example).

That bug report was very helpful, I'm going to try out some of the examples and generally look into this some more. Hopefully there's a simple workaround but I would also be ok with putting a pin in this here and addressing it when we have some of the higher level objects and can measure the impact.

emkornfield replied:
Yeah, I don't know that there is a great solution for float, so it is ok to bypass it.  I think double should exactly round-trip.  If it doesn't I need to further refine my understanding.

samredai replied:
Oh right, yes double does exactly round trip. I moved the double test out of the approximation test function to the equality test function.

emkornfield commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,288 @@
+        (DecimalType(29, 24), b"'\xe3\xa5\x93\x92\x97\xa4\xc7W\xd8\x11\x80", Decimal("12345.123456789123456789123456")),
+        (DecimalType(30, 25), b"\x01\x8e\xe4w\xc3\xb9\xeco\xc9np\xaf\x07", Decimal("12345.1234567891234567891234567")),
+        (DecimalType(31, 26), b"\x0f\x94\xec\xad\xa5C<]\xdePf\xd6N", Decimal("12345.12345678912345678912345678")),
+    ],
```
It would be nice to test Min/Max values that a particular precision/scale can hold.  i.e. "99999.99999", "-99999.99999"

samredai replied:
Great idea, I'll add that in.

samredai replied:
Added `test_max_value_round_trip_conversion` and `test_min_value_round_trip_conversion`.

emkornfield commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,324 @@
+
+
+@pytest.mark.parametrize(
+    "primitive_type, expected_max_value",
```
nit: min_value here

emkornfield replied:
and for parameter name.

samredai replied:
Ah thanks for catching that, fixed!

samredai commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,288 @@
+@pytest.mark.parametrize(
+    "primitive_type, b, approximate_result, approximation",
+    [
+        (FloatType(), b"\x19\x04\x9e?", 1.2345, 5),
```
Good idea, I added a comment here and also in the `from_bytes(...)` function that's dispatched for `FloatType`. Let me know if it looks accurate and clear enough for if we have to revisit this in the future.

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+@from_bytes.register(FloatType)
+def _(primitive_type, b: bytes):
+    """
+    Note: float in python is implemented using a double in C. Therefore this involves a conversion of a 32-bit (single precision)
```
I think this comment belongs on to_bytes, since the actual precision loss is going from double->float.  Float can be converted without loss to double.

samredai replied:
Ah right, that makes sense. Ok I'll move this there!

samredai replied:
Thinking about this some more, should this be described in both `to_bytes` and `from_bytes`? My understanding is that although the precision is lost when going `python float(64-bit) -> c float(32-bit)`, there's some imprecision that happens when going `c float(32-bit) -> python float(64-bit)`.
```c
int main(void) {
  int i;
  union {
    float f;
    unsigned char bytes[4];
  } packed;

  packed.f = 1.234;
  for (i=0; i<4; i++)
    printf ("\\x%02x", packed.bytes[i]);
  printf("\n");
}
```
output:
```
\xb6\xf3\x9d\x3f
```
```py
import struct

print(struct.unpack("<f", b"\xb6\xf3\x9d\x3f")[0])
```
output:
```
1.2339999675750732
```

emkornfield replied:
I guess it depends on your definition of precision. This arises because inability to actually represent 1.234 in ieee floats, so the  approximation differs between types.  This [stack-overflow post](https://stackoverflow.com/questions/17504833/why-converting-from-float-to-double-changes-the-value) I think gives a decent explanation.

```
>>> f"{1.2339999675750732:f}"
'1.234000'
>>> float(f"{1.2339999675750732:f}")
1.234
```

The only actual differences that I'm aware of between float and double is the bit-width for mantissa and exponent, both of which are strictly greater for double so the conversion should always be lossless, so in this case it would be precise.  How humans interpret that lossless conversion is interesting, and probably relevant for how filters are handled.

Printing doubles has some [research behind it](https://cseweb.ucsd.edu/~lerner/papers/fp-printing-popl16.pdf).  This is also why when people want to track currency, they use Decimal and not floats.

samredai replied:
Great, thanks for the explanation. That makes sense!

samredai replied:
Moved the note to `to_bytes`.

cccs-eric commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,329 @@
+        (FixedType(3), b"foo", b"foo"),
+        (BinaryType(), b"foo", b"foo"),
+        (DecimalType(5, 2), b"\x30\x39", Decimal("123.45")),
+        (DecimalType(7, 4), b"\x00\x12\xd6\x87", Decimal("123.4567")),
```
Suggestion: when I migrated the Java tests to python_legacy, I really appreciated the test comments as it helped me understand what I was actually testing.  For some tests, it might not always be obvious.  For example, the test with `Decimal("123.4567")` is not a random value, but a value that stored on 3 bytes, which is not a power of 2 (which is usually used for int/long, etc...).  The Iceberg spec states to use the minimum number of bytes to store the value.  

It doesn't add value to the code, but it can be very helpful for a new dev when things go sideways...  Again, just a suggestion since I found it useful when I had to play in that code.

samredai replied:
This is a great catch, I'll update this PR by bringing over those test comments and maybe add some additional ones. Thanks @cccs-eric !

samredai replied:
I brought over all of the comments and organized them in a module-level docstring in `test_conversions.py`. I also brought over additional tests from the legacy library.

cccs-eric replied:
Nicely done.  Looks pro! 😛

cccs-eric reviewed (approved):
LGTM 👍🏻

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+@from_partition_value_to_py.register(IntegerType)
+@from_partition_value_to_py.register(LongType)
+def _(primitive_type, partition_value_as_str: str):
+    return int(float(partition_value_as_str))
```
Do these also need checks for `None`?

rdblue replied:
Converting to float and then to int allows values through that would fail in Java.

I think we want to be strict here as well. If a value has a decimal point and isn't an integer, I think it is better to fail. This is only used in narrow cases anyway. In fact, I thought about saying we should remove these methods entirely. Let's leave them in, but be strict about the conversions.

samredai replied:
Thanks for catching this! I added some logic to raise if there are any non-zero fractional digits. ([here](https://github.com/apache/iceberg/pull/4263/files#diff-23863ae95c4894ac0025efabd97dd925503dc5a908e386d7c345952e36b9743cR134)):
```py
@from_partition_value_to_py.register(IntegerType)
@from_partition_value_to_py.register(LongType)
def _(primitive_type, partition_value_as_str: str):
    """
    Note: Before attempting to cast the value to string, a validation happens to ensure that there are no fractional digits. For
    example, an invalid value such as "123.45" will be converted to a decimal with digits (1, 2, 3, 4, 5) with an exponent of -2 and the last
    2 digits will be inspected for non-zero values. An example of a valid value is "123.00".
    """
    if partition_value_as_str is None:
        return None
    _, digits, exponent = Decimal(partition_value_as_str).as_tuple()
    if exponent != 0 and any(digits[exponent:]):  # If there are digits to the right of the exponent, raise if any are not 0
        raise ValueError(f"Cannot convert partition value, value cannot have fractional digits for {primitive_type} partition")
    return int(float(partition_value_as_str))
```

rdblue replied:
I think just directly casting to int is a simpler way to do this. When I remove the inner `float` cast and test it, I get `ValueError: invalid literal for int() with base 10: '12.34'`. That seems correct to me. This method is also _not_ recommended for production use, it is more of an internal convenience.

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+
+@from_partition_value_to_py.register(FixedType)
+def _(primitive_type, partition_value_as_str: str):
+    return bytearray(bytes(partition_value_as_str, "UTF-8"))
```
I think we will use `bytes` instead of `bytearray`. So we should combine this with `BinaryType` like `IntType` and `LongType`.

rdblue replied:
The `to_bytes` conversion below also uses bytes, so that will be more consistent as well.

samredai replied:
Fixed [here](https://github.com/apache/iceberg/pull/4263/files#diff-23863ae95c4894ac0025efabd97dd925503dc5a908e386d7c345952e36b9743cR164), thanks!

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+
+@singledispatch
+def to_bytes(primitive_type, value) -> bytes:
+    """A generic function which converts a built-in python value to bytes
```
I think it would be good to refer to the spec here, since these implement single-value serialization: https://iceberg.apache.org/spec/#appendix-d-single-value-serialization

samredai replied:
Great idea, added this to the docstring:
```
This conversion follows the serialization scheme for storing single values as individual binary values defined in the Iceberg specification that
can be found at https://iceberg.apache.org/spec/#appendix-d-single-value-serialization
```

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+
+
+@singledispatch
+def from_partition_value_to_py(primitive_type, partition_value_as_str: str):
```
Should this use `PrimitiveType` for the type argument? Or does that not matter because it uses `@singledispatch`?

samredai replied:
Hmm, I think in some cases it may not matter but in this case since all of the registered functions inherit from the same superclass I think it's a very useful type-hint. Added it [here](https://github.com/apache/iceberg/pull/4263/files#diff-23863ae95c4894ac0025efabd97dd925503dc5a908e386d7c345952e36b9743cR175).

I also added a type-hint for the `value` argument that describes an aggregate of all of the types possible for the registered functions. Another thing is I noticed I was missing a lot of return type-hints so I added those as well. 👍

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+        )
+
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
```
I think this needs to be `primitive_type.precision + primitive_type.scale`.

Say I have a decimal(4,2) value, 12.34. The unscaled value is 12.34 with precision 4. After multiplying by 10**2, the intermediate decimal value becomes 1234.00, which precision 6. So if the context's precision is limited then that could be a problem because the multiplication temporarily creates a value outside the precision.

emkornfield replied:
I don't think this is the case.  I think precision here refers to significant digits sense, not in the DB sense (probably worth a comment):
```
>>> a = Decimal("123.45")
>>> with localcontext() as ctx:
...   ctx.prec = 1
...   a * 10 ** 2
... 
Decimal('1E+4')

>>> with localcontext() as ctx:
...   ctx.prec = 5
...   a * 10 ** 2
... 
Decimal('12345')
```

emkornfield replied:
probably adding a comment to avoid confusion makes sense.

samredai replied:
I agree with @emkornfield in this particular case no precision is lost since the significant figures aren't changing. However I think adjusting the scale in a more direct way avoids this problem entirely and shouldn't require any additional notes. This method recreates the decimal using the same sign and digits and just adjusts the exponent/scale to the new value.
```py
    with localcontext() as ctx:
        ctx.prec = decimal_type_var.precision
        sign, digits, exponent = value.as_tuple()
        value_w_adjusted_scale = Decimal((sign, digits, exponent + decimal_type_var.scale))
        return int(value_w_adjusted_scale.to_integral_value())
```
Let me know what ya'll think.

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+
+
+@to_bytes.register(DecimalType)
+def _(primitive_type, value: Decimal) -> bytes:
```
Other than the precision issue below, I think that this is a correct conversion. We may want to break this up into other decimal functions for use elsewhere. For example, getting the unscaled value might be useful for the truncate transform. Also, the hash transform should call this method! (FYI @jun-he)

rdblue replied:
Actually, I think the min_num_bytes calculation is incorrect.

samredai replied:
Broke this out into a `convert_decimal_to_unscaled_value(...)` method at the top of the module.

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+
+
+@singledispatch
+def from_bytes(primitive_type, b: bytes):
```
Minor: Does Python encourage short variable names like `b`? I think most languages discourage single-character names.

samredai replied:
In general no but I think `b` for bytes and `f` for file are the two exceptions. Also `bytes` is a reserved keyword, I'm not opposed to changing this though, is there a better name here?

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+
+@from_bytes.register(UUIDType)
+def _(primitive_type, b: bytes):
+    return uuid.UUID(int=struct.unpack(">QQ", b)[0] << 64 | struct.unpack(">QQ", b)[1])
```
This runs `unpack` on the bytes twice. It would be better to use a separate variable to avoid doing the work twice.

samredai replied:
Ah I see, updated it to avoid the repeated unpack. Thanks! ([here](https://github.com/apache/iceberg/pull/4263/files#diff-23863ae95c4894ac0025efabd97dd925503dc5a908e386d7c345952e36b9743cR292))

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+    integer_representation = int.from_bytes(b, "big", signed=True)
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        return Decimal(integer_representation) * (Decimal(10) ** Decimal(-primitive_type.scale))
```
I think this should construct the multiplier like the [handling in the `literals.py` PR](https://github.com/apache/iceberg/pull/4262/files#r832771258). That's a more direct way to create the `0.001` value:

```python
Decimal(integer_representation) * Decimal((0, (1,), -primitive_type.scale))
```

samredai replied:
Updated [here](https://github.com/apache/iceberg/pull/4263/files#diff-23863ae95c4894ac0025efabd97dd925503dc5a908e386d7c345952e36b9743cR307).

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,422 @@
+        (UUIDType(), "f79c3e09-677c-4bbd-a479-3f349cb785e7", uuid.UUID("f79c3e09-677c-4bbd-a479-3f349cb785e7")),
+        (FixedType(3), "foo", bytearray(b"foo")),
+        (BinaryType(), "foo", b"foo"),
+        (None, None, None),
```
You may want a separate test that validates for all primitive types, None produces None.

samredai replied:
Add a [test_none_partition_values](https://github.com/apache/iceberg/pull/4263/files#diff-704ab9ed1661567f3b0eb10979903d6a9eb69efa2a76d099ed517ca19e5889beR143) test and it highlighted that a few weren't returning None for None so I updated those functions as well.

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,422 @@
+        (DecimalType(5, 2), "123.45", Decimal("123.45")),
+        (StringType(), "foo", "foo"),
+        (UUIDType(), "f79c3e09-677c-4bbd-a479-3f349cb785e7", uuid.UUID("f79c3e09-677c-4bbd-a479-3f349cb785e7")),
+        (FixedType(3), "foo", bytearray(b"foo")),
```
This should be bytes

samredai replied:
Fixed!

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,422 @@
+        (BinaryType(), bytearray("Z", "utf8"), bytes([90])),
+        (BinaryType(), b"foo", b"foo"),
+        (DecimalType(5, 2), b"\x30\x39", Decimal("123.45")),
+        (DecimalType(7, 4), b"\x00\x12\xd6\x87", Decimal("123.4567")),
```
I don't think this is correct. The number of bytes should be 3, not 4, because null bytes are omitted.

The calculation for min bytes is:
```python
min_num_bytes = ((unscaled_value).bit_length() + 7) // 8
```

That translates to:
```python
((1234567).bit_length() + 7) // 8
=> 3
```

So the calculation looks correct. I'm wondering why the byte array ends up with the extra null byte at the start.

emkornfield replied:
This is just a test issue.  This test is from_bytes (it isn't verifying round-trip) so the extra bytes don't impact the tests because they the extra padding can still be converted.  The to_bytes tests look like they have the correct lengths.

samredai replied:
Thanks everyone, I removed the extra byte and also changed the round trip to use bytes instead of bytearray just for consistency.

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,422 @@
+    using a double in C. Therefore a 32-bit (single precision) float is unpacked to 64-bit (double precision)
+    float in python which introduces some imprecision.
+    """
+    assert conversions.from_bytes(primitive_type, b) == pytest.approx(approximate_result, approximation)
```
I don't think this should be an approximate result. Instead, can you send the expected result through pack/unpack to produce the same 32-bit value?

samredai replied:
Ah right, added the pack/unpack and it works without any approximation.

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+@from_partition_value_to_py.register(BooleanType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return False
```
In this case since it is very constrained values it could be but the general recommendation is to use [direct comparisons with None](https://peps.python.org/pep-0008/#programming-recommendations) if that is actually your intent.

samredai replied:
I made an assumption here that `None` is all we need to explicitly check for. Thinking about this more for the other functions, how do we want to handle a `StringType()` or `IntegerType()` partition that has a partition value of an empty string? Should that also return `None`?

emkornfield commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,267 @@
+
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        unscaled_value = int((value * 10**primitive_type.scale).to_integral_value(rounding=ROUND_HALF_DOWN))
```
nit, I don't think rounding should matter here since there shouldn't be anything after the decimal point?

samredai replied:
Good point, removed the rounding. Thanks!

samredai commented:
@rdblue @emkornfield I appreciate the reviews! I think I addressed everything and I also rebased with the changes from the recently merged doctests PR.

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,294 @@
+def to_bytes(primitive_type: PrimitiveType, value: Union[bool, bytes, Decimal, float, int, str, uuid.UUID]) -> bytes:
+    """A generic function which converts a built-in python value to bytes
+
+    This conversion follows the serialization scheme for storing single values as individual binary values defined in the Iceberg specification that
```
Do we need to wrap lines in docstrings like this? It seems long?

samredai replied:
Good catch I would have expected the auto-formatter to chop this line. Fixed!

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,294 @@
+    with localcontext() as ctx:
+        ctx.prec = decimal_type_var.precision
+        sign, digits, exponent = value.as_tuple()
+        value_w_adjusted_scale = Decimal((sign, digits, exponent + decimal_type_var.scale))
```
When converting using the direct tuple construction, do we need to adjust the local context? According to [this intro to Decimal page](https://www.pythontutorial.net/advanced-python/python-decimal/), "Notice that the decimal context precision only affects the arithmetic operation, not the Decimal constructor".

That means we should just be able to return the adjusted value with scale 0.

Also, since we don't need to set the context precision, we don't need the decimal type at all. In fact, it doesn't make much sense to check the value against the type here either. I think this method could be focused on just getting the unscaled value and the checks can happen in the caller, if needed.

samredai replied:
Oh wow you're absolutely right. I made those changes and it's much cleaner now.

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,294 @@
+)
+
+
+def convert_decimal_to_unscaled_value(decimal_type_var: DecimalType, value: Decimal) -> int:
```
What about a bit simpler name? This could be `decimal_to_unscaled`. The "convert" is implied by being in `conversions` and "value" is unnecessary because this clearly returns a value.

samredai replied:
Haha yeah that was a pretty ugly function name 😅 . Updated!

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,294 @@
+    integer_representation = int.from_bytes(b, "big", signed=True)
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        return Decimal(integer_representation) * Decimal((0, (1,), -primitive_type.scale))
```
One more idea here: `Decimal(integer_representation)` creates a decimal with the right sign and digits. The right scale is `-primitive_type`. So we can avoid doing any decimal arithmetic (and hitting context problems) by using this:

```python
def unscaled_to_decimal(unscaled: int, scale: int) -> Decimal:
    sign, digits, _ = Decimal(unscaled).to_tuple()
    return Decimal((sign, digits, scale))

@from_bytes.register(DecimalType)
def _(decimal_type, buf: bytes) -> Decimal:
    unscaled = int.from_bytes(buf, "big", signed=True)
    return unscaled_to_decimal(unscaled, -primitive_type.exponent)
```

samredai replied:
Much cleaner! Switched to this and removed `localcontext` use entirely.

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,294 @@
+@from_partition_value_to_py.register(BooleanType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return None
```
Instead of adding this check in every method, could we use an outer function that calls the `@singledispatch` method?

```python
def partition_to_py(primitive_type: PrimitiveType, value_str: str):
    if value_str is None:
        return None
    elif value_str == '__HIVE_DEFAULT_PARTITION__':
        return None
   return _partition_to_py(primitive_type, value_str)

@singledispatch
def _partition_to_py(primitive_type: PrimitiveType, value_str: str):
    raise TypeError("Cannot convert '{value_str}' to unsupported type: {primitive_type}")
```

samredai replied:
Another option is to create a simple `@handle_none` decorator and decorate the relevant functions. I think this may be a bit better when navigating the code in the future. When using a "Go to Definition" somehwere `partition_to_py` is used, it will take you straight to the definition without the double hop and I think most people would ignore the `@handle_none` decorator unless it sounds relevant to what they're looking for. (see here and let me know what you think)
```py
def handle_none(func):
    """A decorator function to handle cases where partition values are `None` or "__HIVE_DEFAULT_PARTITION__"

    Args:
        func (Callable): A function registered to the singledispatch function `partition_to_py`
    """

    def wrapper(primitive_type, partition_value_as_str):
        if partition_value_as_str is None:
            return None
        elif partition_value_as_str == "__HIVE_DEFAULT_PARTITION__":
            return None
        return func(primitive_type, partition_value_as_str)

    return wrapper


@partition_to_py.register(BooleanType)
@handle_none
def _(primitive_type, value_str: str) -> Union[int, float, str, uuid.UUID]:
    return value_str.lower() == "true"
```

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,442 @@
+        (LongType(), b"\xd2\x04\x00\x00\x00\x00\x00\x00", 1234),
+        (DoubleType(), b"\x8d\x97\x6e\x12\x83\xc0\xf3\x3f", 1.2345),
+        (DateType(), bytes([232, 3, 0, 0]), 1000),
+        (DateType(), b"\xd2\x04\x00\x00", 1234),
```
What is the purpose of testing both `bytes` representations? Is this to match existing tests in other places?

samredai replied:
Replaced these with byte strings. I don't think there's any purpose for having both representations or even the tests that validate that the resulting strings are utf-8, these were brought over from the legacy library. I think they may have been relevant at the time because python 2 was still being supported which had some different encoding behaviors compared to python 3+.

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,442 @@
+
+
+@pytest.mark.parametrize(
+    "primitive_type, expected_min_value",
```
These don't look like expected min values? They match the expected max cases.

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,442 @@
+@pytest.mark.parametrize(
+    "primitive_type, value, expected_error_message",
+    [
+        (DecimalType(7, 3), Decimal("123.4567"), "Cannot serialize value, scale of value does not match type decimal(7, 3): 4"),
```
We can probably omit a few words from the error message here: "scale does not match {decimal_type}"

samredai replied:
Fixed!

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,442 @@
+        (LongType(), "123.45", True),
+        (LongType(), "1234567.89", True),
+        (LongType(), "123.00", False),
+        (LongType(), "1234567.00", False),
```
I think these should throw exceptions as well. If there is a decimal point, then the type does not match the original type and it should fail.

samredai replied:
Ah got it, I misunderstood that but it makes sense. Fixed to always raise if there's a decimal point and added some sets that make sure it does not raise when there is no decimal point (as a sanity-check).

rdblue commented on python/tests/test_conversions.py:
```diff
@@ -0,0 +1,442 @@
+        (LongType(), "1234567.00", False),
+    ],
+)
+def test_from_partition_value_to_py_raise_on_incorrect_precision_or_scale(primitive_type, value, should_raise):
```
Can you relocate the partition tests so they're grouped together?

samredai replied:
Done! 👍

samredai commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,294 @@
+
+@from_bytes.register(DecimalType)
+def _(primitive_type, b: bytes) -> Decimal:
+    integer_representation = int.from_bytes(b, "big", signed=True)
```
Sounds good! Updated

samredai commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,318 @@
+    Returns:
+        int: The unscaled value
+    """
+    return int(value.to_integral_value())
```
Thanks for catching that. Pushed a fix and also added `test_decimal_to_unscaled` and `test_unscaled_to_decimal` to test calling those functions directly.

rdblue commented on python/src/iceberg/conversions.py:
```diff
@@ -0,0 +1,319 @@
+            f"Cannot serialize value, precision of value is greater than precision of type {primitive_type}: {len(digits)}"
+        )
+
+    unscaled_value = decimal_to_unscaled(value=Decimal((sign, digits, 0)))
```
Minor: this could pass `value` now that `decimal_to_unscaled` is fixed.

rdblue commented:
Thanks, @samredai!
